[JSC] Add new internal microtask `PromiseFirstResolveWithoutHandlerJob` for `Promise.race`

sosukesuzuki opened the pull request:
#### ad5c326d784521842d7b246bdd546f8ef236c85b
<pre>
[JSC] Add new internal microtask `PromiseFirstResolveWithoutHandlerJob` for `Promise.race`
<a href="https://bugs.webkit.org/show_bug.cgi?id=302151">https://bugs.webkit.org/show_bug.cgi?id=302151</a>

Reviewed by Yusuke Suzuki.

This patch adds a new internal microtask, `PromiseFirstResolveWithoutHandlerJob`.

It is very similar to the existing `PromiseResolveWithoutHandlerJob`, but differs
in that it returns early without doing anything if the target promise is already
fulfilled.

This reduces unnecessary function allocations in the fast path of `Promise.race`.

                        TipOfTree                  Patched

promise-race          7.4727+-0.3047     ^      3.4886+-0.1515        ^ definitely 2.1420x faster

* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runInternalMicrotask):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::performPromiseThenWithInternalMicrotask):
(JSC::JSPromise::triggerPromiseReactions):
* Source/JavaScriptCore/runtime/JSPromise.h:
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/Microtask.h:

Canonical link: <a href="https://commits.webkit.org/302896@main">https://commits.webkit.org/302896@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee1227a8ac0bdd0b49ad2c0a0b9609e363d4df04

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130570 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2841 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41525 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137988 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2853 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2733 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99475 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133517 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116923 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80181 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35049 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81247 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122573 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/110563 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35555 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140467 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129023 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2631 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2395 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2675 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113263 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107911 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27461 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2034 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31694 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55609 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2701 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66090 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162038 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2520 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40406 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/2722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2627 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -285,6 +286,30 @@ void JSPromise::performPromiseThen(VM& vm, JSGlobalObject* globalObject, JSValue
+    JSValue reactionsOrResult = this->reactionsOrResult();
+    switch (status()) {
+    case JSPromise::Status::Pending: {
+        auto* context = JSPromiseRaceContext::create(vm, promise);
```
Why do we need to have new context? Why not passing `promise` as a context?

Constellation replied:
If you encode `InternalMicrotask` into `handler`, then you do not need to have this object.

Constellation commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -538,7 +563,8 @@ void JSPromise::triggerPromiseReactions(VM& vm, JSGlobalObject* globalObject, St
         if (handler.isUndefinedOrNull()) {
-            globalObject->queueMicrotask(InternalMicrotask::PromiseResolveWithoutHandlerJob, promise, argument, jsNumber(static_cast<int32_t>(status)), jsUndefined());
+            bool isFirstResolveOnly = !context.isEmpty() && context.isCell() && context.asCell()->type() == JSPromiseRaceContextType;
+            globalObject->queueMicrotask(isFirstResolveOnly ? InternalMicrotask::PromiseFirstResolveWithoutHandlerJob : InternalMicrotask::PromiseResolveWithoutHandlerJob, promise, argument, jsNumber(static_cast<int32_t>(status)), jsUndefined());
```
This looks too costly. Note that this code is really hot, so you cannot make this code more costly (accessing heap etc.).

I think better approach would be putting Int32 to handler and holding `InternalMicrotask` value in `handler`.

sosukesuzuki replied:
Thanks, fixed!

Constellation commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -285,6 +285,30 @@ void JSPromise::performPromiseThen(VM& vm, JSGlobalObject* globalObject, JSValue
     markAsHandled();
 }
 
+void JSPromise::performPromiseThenForPromiseRace(VM& vm, JSGlobalObject* globalObject, JSValue promise)
```
We do not need to limit it for `ForPromiseRace` case. You can extend this interface to take `InternalMicrotask` and then you can define it as `performPromiseThenWithInternalMicrotask`.

Constellation commented on Source/JavaScriptCore/runtime/JSMicrotask.cpp:
```diff
@@ -212,8 +212,11 @@ void runInternalMicrotask(JSGlobalObject* globalObject, InternalMicrotask task,
     case InternalMicrotask::PromiseResolveWithoutHandlerJob: {
         auto* promise = jsCast<JSPromise*>(arguments[0]);
+        if (task == InternalMicrotask::PromiseFirstResolveWithoutHandlerJob && promise->status() != JSPromise::Status::Pending)
+            return;
```
I would like to make it,
```
case InternalMicrotask::PromiseFirstResolveWithoutHandlerJob:
    if (jsCast<JSPromise*>(arguments[0])->status() != JSPromise::Status::Pending)
        return;
    [[fallthrough]];
case InternalMicrotask::PromiseResolveWithoutHandlerJob: {
```

Constellation commented on Source/JavaScriptCore/runtime/JSPromise.cpp:
```diff
@@ -547,6 +571,12 @@ void JSPromise::triggerPromiseReactions(VM& vm, JSGlobalObject* globalObject, St
+            globalObject->queueMicrotask(task, promise, argument, jsNumber(static_cast<int32_t>(status)), jsUndefined());
+            continue;
+        }
+
```
It should be handled before `promise.isUndefinedOrNull()` check. And I think we should unify this and `handler.isUndefinedOrNull()` case. No-handler case should take `InternalMicrotask::PromiseResolveWithoutHandlerJob` for handler instead.

Constellation reviewed (approved):
Nice!

sosukesuzuki committed:
[JSC] Add new internal microtask `PromiseFirstResolveWithoutHandlerJob` for `Promise.race`
https://bugs.webkit.org/show_bug.cgi?id=302151

Reviewed by Yusuke Suzuki.

This patch adds a new internal microtask, `PromiseFirstResolveWithoutHandlerJob`.

It is very similar to the existing `PromiseResolveWithoutHandlerJob`, but differs
in that it returns early without doing anything if the target promise is already
fulfilled.

This reduces unnecessary function allocations in the fast path of `Promise.race`.

                        TipOfTree                  Patched

promise-race          7.4727+-0.3047     ^      3.4886+-0.1515        ^ definitely 2.1420x faster

* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::runInternalMicrotask):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::performPromiseThenWithInternalMicrotask):
(JSC::JSPromise::triggerPromiseReactions):
* Source/JavaScriptCore/runtime/JSPromise.h:
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/Microtask.h:

Canonical link: https://commits.webkit.org/302896@main
